### PR TITLE
Record scopes in the simple and JSON loggers

### LINKED ISSUE #193: JsonConsoleLogger and SimpleConsoleLogger do not support BeginScope
I wrote the following code and was expecting, that the logger will write the scope keys (in my example key, JobId and fileName inside the logging. When I checked the logging on the UI the keys are not written. 

``` csharp
using (this._logger.BeginScope(new Dictionary<string, object> { ["key"] = "1234567890123", ["jobId"] = 54 }))
{
    this._logger.LogTrace("Test Log - Scope Map - Trace");
    this._logger.LogDebug("Test Log - Scope Map - Debug");
    this._logger.LogInformation("Test Log - Scope Map - Information");
    this._logger.LogWarning("Test Log - Scope Map - Warning");
    this._logger.LogError("Test Log - Scope Map - Error");
    this._logger.LogError(new Exception("My Error Exception"), "Test Log - Scope Map - Error with Exception");
    this._logger.LogCritical("Test Log - Scope Map - Critical");
    this._logger.LogCritical(new Exception("My Critical Exception"), "Test Log - Scope Map - Critical with Exception");

    using (this._logger.BeginScope(new Dictionary<string, object> { ["fileName"] = "Test.txt" }))
    {
        this._logger.LogTrace("Test Log - Scope File - Trace");
        this._logger.LogDebug("Test Log - Scope File - Debug");
        this._logger.LogInformation("Test Log - Scope File - Information");
        this._logger.LogWarning("Test Log - Scope File - Warning");
        this._logger.LogError("Test Log - Scope File - Error");
        this._logger.LogError(new Exception("My Error Exception"), "Test Log - Scope File - Error with Exception");
        this._logger.LogCritical("Test Log - Scope File - Critical");
        this._logger.LogCritical(new Exception("My Critical Exception"), "Test Log - Scope File - Critical with Exception");
    }
}
```

Maybe it could be an option to include a third logger in the framework, where an output format string can be set via dependency injection. Inside this format string, the scope properties could be used when they are set. 

Example: "{scope:jobId} {scope:key} {message}"

Alternatively the logger could write the scope to the JsonPayload.

### PR DESCRIPTION
**Pull request by jskeet**
Fixes #193

### CODE REVIEW

File: src/Google.Cloud.Functions.Hosting/Logging/SimpleConsoleLogger.cs [hunk 1]
```diff
@@ -37,12 +42,54 @@ protected override void LogImpl<TState>(LogLevel logLevel, EventId eventId, TSta
...
-            // nice way of getting all the relevant information otherwise.
+            StringBuilder scopeBuilder = new StringBuilder();
+            ScopeProvider.ForEachScope(AppendScope, scopeBuilder);
+
+            _console.WriteLine(Invariant($"{DateTime.UtcNow:yyyy-MM-dd'T'HH:mm:ss.fff'Z'} [{Category}] [{briefLevel}] {formattedMessage}"));
+            // Note: it's not ideal to break out of the "one line per log entry" approach for either scopes or exceptions,
+            // but there's no particularly nice way of getting all the relevant information otherwise.
+            if (scopeBuilder.Length != 0)
+            {
+                _console.WriteLine($"Scopes: [{scopeBuilder}]");
```
Comment (amanda-tarafa) [inline src/Google.Cloud.Functions.Hosting/Logging/SimpleConsoleLogger.cs:53]: I would have expected the scopes to go first I think. But I understand that this makes sense. Maybe add some visual dependency sign like "|__" (ugly, I know)

File: src/Google.Cloud.Functions.Hosting/Logging/JsonConsoleLogger.cs [hunk 2]
```diff
@@ -107,5 +117,36 @@ bool ContainsFormatParameters(IEnumerable<KeyValuePair<string, object>> fields)
...
+        private static void WriteScope(object value, JsonWriter writer)
+        {
+            // Detect "first scope" and start the scopes array property.
+            if (writer.WriteState == WriteState.Object)
+            {
+                writer.WritePropertyName("scopes");
+                writer.WriteStartArray();
+            }
+
+            if (value is IEnumerable<KeyValuePair<string, object>> kvps)
```
Comment (amanda-tarafa) [inline src/Google.Cloud.Functions.Hosting/Logging/JsonConsoleLogger.cs:130]: Same comment as before, maybe just check for an IDictionary?

File: src/Google.Cloud.Functions.Hosting/Extensions/FunctionsFrameworkLoggingExtensions.cs [hunk 3]
```diff
@@ -35,8 +35,8 @@ public static ILoggingBuilder AddFunctionsConsoleLogging(this ILoggingBuilder bu
         {
```
Comment (amanda-tarafa) [inline src/Google.Cloud.Functions.Hosting/Extensions/FunctionsFrameworkLoggingExtensions.cs:35]: nit: Line 28 mentions "single line per log entry". Might not be the best place to go into a lengthy explanation, but maybe a note about scopes and exceptions might be good.

**Review by amanda-tarafa — APPROVED** [review]
LGTM, a couple of comments, but it's also fine as it is.

File: src/Google.Cloud.Functions.Hosting/Logging/SimpleConsoleLogger.cs [hunk 4]
```diff
@@ -37,12 +42,54 @@ protected override void LogImpl<TState>(LogLevel logLevel, EventId eventId, TSta
...
+        }
+
+        private void AppendScope(object value, StringBuilder builder)
+        {
+            if (builder.Length != 0)
+            {
+                builder.Append(", ");
+            }
+            // If the scope is a dictionary, format it as key/value pairs
+            if (value is IEnumerable<KeyValuePair<string, object>> kvps)
```
[the remark this answers is not in the document]
Comment (jskeet) [reply src/Google.Cloud.Functions.Hosting/Logging/SimpleConsoleLogger.cs:69]: Hmm... this is what we do in GoogleLogger, and it's also the recommendation in https://nblumhardt.com/2016/11/ilogger-beginscope/.

File: src/Google.Cloud.Functions.Hosting/Logging/SimpleConsoleLogger.cs (hunk 1 above)
Comment (jskeet) [reply src/Google.Cloud.Functions.Hosting/Logging/SimpleConsoleLogger.cs:53]: Hmmm... I think the lack of a timestamp makes it clear enough, personally - basically the log entry is everything from one timestamp to the next one. It's fairly "busy" already, but I think it's reasonable clear as it is.

File: src/Google.Cloud.Functions.Hosting/Extensions/FunctionsFrameworkLoggingExtensions.cs (hunk 3 above)
Comment (jskeet) [reply src/Google.Cloud.Functions.Hosting/Extensions/FunctionsFrameworkLoggingExtensions.cs:35]: Yup, will do.

**Pull request merged**